cmd/metal-api/internal/service/machine-service.go

```diff
@@ -1493,31 +1524,38 @@ func (r machineResource) finalizeAllocation(request *restful.Request, response *
 }
 
 func (r machineResource) freeMachine(request *restful.Request, response *restful.Response) {
+	r.reinstallOrDeleteMachine(request, response, nil)
+}
+
+// reinstallOrDeleteMachine (re)installs the requested machine with given image by either allocating
+// the machine if not yet allocated or not modifying any other allocation parameter than 'ImageID'
+// and 'Reinstall' set to true.
+// If the given image ID is nil, it deletes the machine instead.
+func (r machineResource) reinstallOrDeleteMachine(request *restful.Request, response *restful.Response, imageID *string) {
 	id := request.PathParameter("id")
 	m, err := r.ds.FindMachineByID(id)
-	if checkError(request, response, utils.CurrentFuncName(), err) {
+	if err != nil {
 		return
 	}
 	logger := utils.Logger(request).Sugar()
 
-	err = freeMachine(r.ds, r, r.ipamer, m, logger)
+	err = reinstallOrDeleteMachine(r.ds, r, r.ipamer, m, imageID, logger)
 	if checkError(request, response, utils.CurrentFuncName(), err) {
 		return
 	}
 
-	err = response.WriteHeaderAndEntity(http.StatusOK, makeMachineResponse(m, r.ds, logger))
+	err = response.WriteHeaderAndEntity(http.StatusOK, makeMachineResponse(m, r.ds, utils.Logger(request).Sugar()))
 	if err != nil {
 		logger.Error("Failed to send response", zap.Error(err))
-		return
 	}
 }
 
-func freeMachine(ds *datastore.RethinkStore, publisher bus.Publisher, ipamer ipam.IPAMer, m *metal.Machine, logger *zap.SugaredLogger) error {
+func reinstallOrDeleteMachine(ds *datastore.RethinkStore, publisher bus.Publisher, ipamer ipam.IPAMer, m *metal.Machine, imageID *string, logger *zap.SugaredLogger) error {
```

Comment:
To me it looks like the original `freeMachine` function and `reinstallMachine` do not share enough code to justify combining them with a more complex function `reinsintallOrDeleteMachine`, adding an additional parameter to the function (completely unnecessary for `freeMachine`) and more conditional branches in the function body. To my taste it would make it a little easier to understand the procedure if there was a separate method `reinstallMachine` and the `freeMachine` method stays untouched. I would also say it is more future proof to separate this functionality right from the beginning. What do you think?